File: models/mlb/cartridges/MLB-M2/snapshot-run.mjs
import fs from 'node:fs/promises'
import path from 'node:path'
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'
import { pathToFileURL } from 'node:url'
import { buildM2Snapshot } from './snapshot.mjs'

const execFileAsync = promisify(execFile)
const rootDir = path.resolve(import.meta.dirname, '..', '..', '..', '..')
const localModelId = path.basename(import.meta.dirname).toUpperCase()
const localCartridgeDir = path.relative(rootDir, import.meta.dirname).replaceAll(path.sep, '/')

const readJson = async (relativePath, fallback = null) => {
  try {
    return JSON.parse(await fs.readFile(path.resolve(rootDir, relativePath), 'utf8'))
  } catch (error) {
    if (error.code === 'ENOENT') return fallback
    throw error
  }
}

const writeJson = async (relativePath, payload) => {
  const absolute = path.resolve(rootDir, relativePath)
  await fs.mkdir(path.dirname(absolute), { recursive: true })
  await fs.writeFile(absolute, `${JSON.stringify(payload, null, 2)}\n`)
}

const parseArgs = () => {
  const args = process.argv.slice(2)
  const options = { date: '', model: process.env.MLB_MODEL_ID || localModelId }
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === '--date') {
      options.date = args[index + 1]
      index += 1
    } else if (args[index] === '--model') {
      options.model = String(args[index + 1] || '').toUpperCase()
      index += 1
    }
  }
  if (!/^\d{4}-\d{2}-\d{2}$/.test(options.date)) throw new Error('Pass --date YYYY-MM-DD')
  return options
}

const runCommand = async (command, args) => {
  try {
    const result = await execFileAsync(command, args, { cwd: rootDir, maxBuffer: 1024 * 1024 * 20 })
    return { ok: true, stdout: result.stdout.trim(), stderr: result.stderr.trim() }
  } catch (error) {
    return {
      ok: false,
      stdout: String(error.stdout || '').trim(),
      stderr: String(error.stderr || error.message || error).trim()
    }
  }
}

const gitInfo = async () => {
  const commit = await runCommand('git', ['rev-parse', 'HEAD'])
  const dirty = await runCommand('git', ['status', '--short'])
  return {
    commit: commit.ok ? commit.stdout : null,
    dirty: dirty.ok ? dirty.stdout.length > 0 : true
  }
}

const listGameFiles = async (date) => {
  const relativeDir = `published-data/slates/${date}/games`
  const absoluteDir = path.resolve(rootDir, relativeDir)
  try {
    return (await fs.readdir(absoluteDir))
      .filter((fileName) => fileName.endsWith('.json') && !fileName.startsWith('rg-'))
      .sort()
      .map((fileName) => `${relativeDir}/${fileName}`)
  } catch (error) {
    if (error.code === 'ENOENT') return []
    throw error
  }
}

const uniqueEntries = (entries) => {
  const byPath = new Map()
  for (const entry of entries) {
    if (entry?.path) byPath.set(entry.path, entry)
  }
  return Array.from(byPath.values()).sort((left, right) => left.path.localeCompare(right.path))
}

const sourceInventory = async () => {
  const registry = await readJson('models/mlb/registry.json', {})
  const manifest = await readJson(`${localCartridgeDir}/manifest.json`, {})
  const rp36 = await readJson('models/mlb/cartridges/MLB-RP36/manifest.json', {})
  const e0 = await readJson('models/mlb/cartridges/MLB-E0/manifest.json', {})
  return uniqueEntries([
    { path: 'models/mlb/registry.json', role: 'model-registry' },
    { path: 'models/registry.json', role: 'top-level-model-registry' },
    { path: `${localCartridgeDir}/manifest.json`, role: 'model-manifest' },
    { path: manifest.entrypoint, role: 'model-runner-wrapper' },
    { path: manifest.outputContract, role: 'output-contract' },
    { path: manifest.modelDescription, role: 'model-description' },
    { path: manifest.modelNotes, role: 'model-notes' },
    { path: manifest.modelLog, role: 'model-log' },
    { path: `${localCartridgeDir}/snapshot.mjs`, role: 'snapshot-builder' },
    { path: `${localCartridgeDir}/verify_snapshot.mjs`, role: 'snapshot-verifier' },
    { path: `${localCartridgeDir}/snapshot-run.mjs`, role: 'run-snapshotter' },
    { path: `${localCartridgeDir}/check_run.mjs`, role: 'run-checker' },
    { path: 'models/mlb/cartridges/MLB-RP36/manifest.json', role: 'relief-addendum-manifest' },
    { path: rp36.entrypoint, role: 'relief-addendum-runner' },
    { path: rp36.outputContract, role: 'relief-addendum-output-contract' },
    { path: rp36.modelDescription, role: 'relief-addendum-description' },
    { path: rp36.modelNotes, role: 'relief-addendum-notes' },
    { path: 'models/mlb/cartridges/MLB-E0/manifest.json', role: 'evaluator-manifest' },
    { path: e0.metricsContract, role: 'metrics-contract' },
    ...(manifest.sourceFiles || []),
    ...(rp36.sourceFiles || []),
    ...((registry.cartridges || []).map((cartridge) => ({
      path: path.posix.join('models/mlb', cartridge.targetPath.replace(/^\.\//, ''), 'manifest.json'),
      role: `${cartridge.role || cartridge.modelId}-manifest`
    })))
  ])
}

const inputInventory = async (date) => uniqueEntries([
  { path: `published-data/slates/${date}/summary.json`, role: 'published-summary' },
  ...(await listGameFiles(date)).map((filePath) => ({ path: filePath, role: 'published-game' })),
  { path: `data-private/predictions/mlb-sides/${date}-veto-artifact.json`, role: 'side-veto-artifact' },
  { path: `data-private/predictions/mlb-player-props/${date}-player-props.json`, role: 'player-props' },
  { path: `data-private/predictions/mlb-home-runs/${date}-statcast-prototype.json`, role: 'home-run-board' },
  { path: `data-private/predictions/mlb-reliever-shadow/${date}-reliever-shadow.json`, role: 'rp36-reliever-shadow' },
  { path: `data-private/lineups/mlb/${date}-lineup-board.json`, role: 'lineup-board' }
])

export const snapshotM2Run = async ({ date, model = process.env.MLB_MODEL_ID || localModelId }) => {
  const registry = await readJson('models/mlb/registry.json', {})
  const active = registry.active || {}
  const modelId = String(model || localModelId).toUpperCase()
  const runId = `mlb-${date}-${active.warehouse || 'MLB-W1'}-${active.features || 'MLB-F0'}-${modelId}-${active.reliefAddendum || 'MLB-RP36'}-${active.evaluator || 'MLB-E0'}`
  const runDir = `data-private/model-runs/mlb/${modelId}/${date}`
  const snapshot = await buildM2Snapshot({ date })
  const sourceFiles = await sourceInventory()
  const inputFiles = await inputInventory(date)

  await writeJson(`${runDir}/snapshot.json`, snapshot)
  const outputTargets = [
    { path: `${runDir}/snapshot.json`, role: 'prediction-snapshot' },
    { path: `${runDir}/run.json`, role: 'run-manifest' }
  ]
  const run = {
    schemaVersion: 1,
    runId,
    sport: 'mlb',
    slateDate: date,
    modelId,
    reliefAddendum: active.reliefAddendum || 'MLB-RP36',
    evaluatorVersion: active.evaluator || 'MLB-E0',
    warehouseVersion: active.warehouse || 'MLB-W1',
    featureVersion: active.features || 'MLB-F0',
    mode: 'pregame',
    status: 'snapshotted',
    snapshottedAt: new Date().toISOString(),
    git: await gitInfo(),
    sourceHash: null,
    inputHash: null,
    outputHash: null,
    snapshotHash: snapshot.snapshotHash,
    artifactSummary: snapshot.artifactSummary,
    sourceFiles: sourceFiles.length,
    inputs: inputFiles.length,
    outputs: outputTargets.length,
    artifacts: outputTargets
  }
  await writeJson(`${runDir}/run.json`, run)

  const indexResult = await runCommand('python3', [
    'models/shared/model-runs/index_runs.py',
    'index',
    '--sport',
    'mlb',
    '--model-id',
    modelId,
    '--date',
    date
  ])
  if (!indexResult.ok) {
    throw new Error(`Failed to index MLB run snapshot in warehouse: ${indexResult.stderr || indexResult.stdout}`)
  }

  return { run, runDir: path.resolve(rootDir, runDir) }
}

const executedUrl = process.argv[1] ? pathToFileURL(path.resolve(process.argv[1])).href : null

if (import.meta.url === executedUrl) {
  const { date, model } = parseArgs()
  const { run, runDir } = await snapshotM2Run({ date, model })
  console.log(JSON.stringify({
    runId: run.runId,
    runDir,
    status: run.status,
    snapshotHash: run.snapshotHash,
    artifactSummary: run.artifactSummary
  }, null, 2))
}

File: models/mlb/cartridges/MLB-RP36/snapshot_run.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parents[4]
MODEL_ID = "MLB-RP36"
MODEL_DIR = ROOT / "models" / "mlb" / "cartridges" / MODEL_ID
RUN_ROOT = ROOT / "data-private" / "model-runs" / "mlb" / MODEL_ID
DB_PATH = ROOT / "data-private" / "warehouse" / "sports.db"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Snapshot an MLB-RP36 reliever addendum run.")
    parser.add_argument("--date", required=True, help="Slate date, YYYY-MM-DD.")
    return parser.parse_args()


def read_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def write_json(path: Path, payload: Any) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(f"{json.dumps(payload, indent=2, sort_keys=False)}\n", encoding="utf-8")


def git_info() -> dict[str, Any]:
    def run(args: list[str]) -> str | None:
        try:
            return subprocess.check_output(args, cwd=ROOT, text=True).strip()
        except Exception:
            return None

    status = run(["git", "status", "--short"]) or ""
    return {
        "commit": run(["git", "rev-parse", "HEAD"]),
        "dirty": bool(status),
    }


def source_inventory() -> list[dict[str, str]]:
    manifest = read_json(MODEL_DIR / "manifest.json")
    entries = [
        {"role": "top-level-model-registry", "path": "models/registry.json"},
        {"role": "mlb-model-registry", "path": "models/mlb/registry.json"},
        {"role": "rp36-manifest", "path": "models/mlb/cartridges/MLB-RP36/manifest.json"},
        {"role": "rp36-output-contract", "path": manifest.get("outputContract")},
        {"role": "rp36-model-description", "path": manifest.get("modelDescription")},
        {"role": "rp36-model-notes", "path": manifest.get("modelNotes")},
        {"role": "rp36-run-snapshotter", "path": "models/mlb/cartridges/MLB-RP36/snapshot_run.py"},
        {"role": "rp36-run-checker", "path": "models/mlb/cartridges/MLB-RP36/check_run.py"},
    ]
    entries.extend(manifest.get("sourceFiles") or [])
    by_path: dict[str, dict[str, str]] = {}
    for entry in entries:
        path = entry.get("path")
        if path:
            by_path[path] = {"role": entry.get("role", "source"), "path": path}
    return sorted(by_path.values(), key=lambda entry: entry["path"])


def sqlite_scalar(query: str, args: tuple[Any, ...]) -> list[dict[str, Any]]:
    import sqlite3

    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    try:
        rows = conn.execute(query, args).fetchall()
        return [dict(row) for row in rows]
    finally:
        conn.close()


def input_inventory(date: str) -> list[dict[str, Any]]:
    queries = [
        (
            "games_today",
            """
            select count(*) as rows, coalesce(sum(game_pk), 0) as game_pk_sum
            from mlb_games
            where game_date = ?
            """,
            (date,),
        ),
        (
            "bullpen_usage_today",
            """
            select count(*) as rows, coalesce(sum(pitcher_id), 0) as pitcher_id_sum
            from mlb_bullpen_usage
            where as_of_date = ?
            """,
            (date,),
        ),
        (
            "bullpen_shape_today",
            """
            select count(*) as rows
            from mlb_team_bullpen_shape_daily
            where as_of_date = ?
            """,
            (date,),
        ),
        (
            "starter_logs_through_date",
            """
            select count(*) as rows, coalesce(sum(pitcher_id), 0) as pitcher_id_sum
            from mlb_starting_pitcher_game_logs
            where game_date <= ?
            """,
            (date,),
        ),
        (
            "reliever_appearances_before_date",
            """
            select count(*) as rows, coalesce(sum(pitcher_id), 0) as pitcher_id_sum,
                   coalesce(sum(pitches_thrown), 0) as pitches_sum
            from mlb_pitcher_appearances
            where game_date < ?
              and pitcher_role = 'reliever'
            """,
            (date,),
        ),
        (
            "plate_appearances_before_date",
            """
            select count(*) as rows, coalesce(sum(game_pk), 0) as game_pk_sum
            from mlb_plate_appearances
            where game_date < ?
            """,
            (date,),
        ),
        (
            "lineup_shape_history",
            """
            select count(*) as rows
            from mlb_lineup_conversion_shape_daily
            where as_of_date <= ?
            """,
            (date,),
        ),
        (
            "lineup_dependency_history",
            """
            select count(*) as rows
            from mlb_lineup_dependency_profiles
            where as_of_date <= ?
            """,
            (date,),
        ),
    ]

    rows = []
    for role, query, args in queries:
        result = sqlite_scalar(query, args)
        rows.append({
            "role": role,
            "query": " ".join(query.split()),
            "rows": result,
            "exists": True,
        })
    return rows


def artifact_paths(date: str) -> list[dict[str, str]]:
    return [
        {
            "role": "reliever-shadow-json",
            "path": f"data-private/predictions/mlb-reliever-shadow/{date}-reliever-shadow.json",
        },
        {
            "role": "reliever-shadow-module",
            "path": f"web/src/lib/day-{date}-reliever-shadow.js",
        },
        {
            "role": "reliever-shadow-report",
            "path": "development-docs/mlb/research/mlb-first-up-reliever-shadow-board-053026.md",
        },
    ]


def build_snapshot(date: str) -> dict[str, Any]:
    artifact_path = ROOT / f"data-private/predictions/mlb-reliever-shadow/{date}-reliever-shadow.json"
    artifact = read_json(artifact_path)
    teams = artifact.get("relieverShadowByTeam") or {}
    return {
        "schemaVersion": 1,
        "modelId": MODEL_ID,
        "date": date,
        "meta": artifact.get("meta") or {},
        "artifactSummary": {
            "relieverTeams": len(teams),
            "candidateCount": (artifact.get("meta") or {}).get("candidateCount"),
        },
    }


def main() -> int:
    args = parse_args()
    date = args.date
    run_dir = RUN_ROOT / date
    run_id = f"mlb-{date}-MLB-RP36"

    source_files = source_inventory()
    input_rows = input_inventory(date)
    snapshot = build_snapshot(date)

    write_json(run_dir / "snapshot.json", snapshot)
    output_targets = [
        *artifact_paths(date),
        {"role": "prediction-snapshot", "path": f"data-private/model-runs/mlb/MLB-RP36/{date}/snapshot.json"},
        {"role": "run-manifest", "path": f"data-private/model-runs/mlb/MLB-RP36/{date}/run.json"},
    ]

    run = {
        "schemaVersion": 1,
        "runId": run_id,
        "sport": "mlb",
        "slateDate": date,
        "modelId": MODEL_ID,
        "mode": "relief-addendum",
        "status": "snapshotted",
        "snapshottedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "git": git_info(),
        "sourceHash": None,
        "inputHash": None,
        "outputHash": None,
        "artifactHash": None,
        "artifactSummary": snapshot["artifactSummary"],
        "sourceFiles": len(source_files),
        "inputs": len(input_rows),
        "outputs": len(output_targets),
        "artifacts": output_targets,
    }
    write_json(run_dir / "run.json", run)

    subprocess.run(
        [
            sys.executable,
            str(ROOT / "models" / "shared" / "model-runs" / "index_runs.py"),
            "index",
            "--sport",
            "mlb",
            "--model-id",
            MODEL_ID,
            "--date",
            date,
        ],
        cwd=ROOT,
        check=True,
    )

    print(json.dumps({
        "runId": run_id,
        "runDir": str(run_dir),
        "status": "snapshotted",
        "artifactHash": None,
        "artifactSummary": snapshot["artifactSummary"],
    }, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
